hal/armv7m/mpu.c

```diff
@@ -0,0 +1,333 @@
+			name, region, "", "all", srBase, "", srBase + (1ul << sizeBit), "");
+		mpu_attrPrint(attr, (rasr & 1));
+	}
+}
+
+
+/* Mark region valid, but disabled */
+static void mpu_regionDisable(u8 idx)
+{
+	if (idx < mpu_common.regMax) {
+		mpu_common.hal.mpu.region[idx].rbar =
+			((mpu_common.hal.mpu.region[idx].rbar | (1ul << 4)) & ~0xful) | (idx & 0xful);
+		mpu_common.hal.mpu.region[idx].rasr &= ~1;
+	}
+}
+
+
+/* Setup single MPU region entry in a local MPU context */
+static int mpu_regionSet(u8 idx, u32 addr, u8 sizeBit, u32 attr, u8 enable, u8 srdMask)
+{
+	u8 tex = 0; /* FIXME: TEX is no used currently, defaults to 0 */
+
+	u8 ap = 1; /* unprivileged no access */
+
+	if (sizeBit < 5 || idx >= mpu_common.regMax)
+		return -1;
+
+	if (attr & mAttrRead)
+		ap = 2; /* unprivileged read only access */
+
+	if (attr & mAttrWrite)
+		ap = 3; /* unprivileged read and write access */
+
+	/* privileged access have read-write access */
+
+	/* clang-format off */
+
+	mpu_common.hal.mpu.region[idx].rbar =
+		(addr & (0x7fffffful << 5))
+		| (1ul << 4)
+		| (idx & 0xful);
+
+	mpu_common.hal.mpu.region[idx].rasr =
+		((((attr & mAttrExec) == 0) & 1ul) << 28)
+		| ((ap & 0x7ul) << 24)
+		| ((tex & 0x7ul) << 19)
+		| ((((attr & mAttrShareable) != 0) & 1ul) << 18)
+		| ((((attr & mAttrCacheable) != 0) & 1ul) << 17)
+		| ((((attr & mAttrBufferable) != 0) & 1ul) << 16)
+		| ((srdMask & 0xfful) << 8)
+		| (((sizeBit - 1) & 0x1ful ) << 1)
+		| (enable != 0);
+
+	/* clang-format on */
+
+	return 0;
+}
+
+
+/* Find best alignment for map and setup a region */
+static int mpu_regionBestFit(u8 region, u32 addr, u32 size, u32 attr, u8 enable, u32 *allocSize)
+{
+	u32 srBase, srSize, alignMask;
+
+	u8 alignBits = addr == 0 ? 32 : _ctz(addr);
+	u8 srdMask = 0, bit = _ctz(size);
+
+	if (bit < alignBits)
+		alignBits = bit;
+
+	alignMask = alignBits < 32 ? ~((1u << alignBits) - 1) : 0;
+
+	if (alignBits < 5)
+		*allocSize = 1u << alignBits;
+	else {
+		alignBits += 3;
+
+		if (alignBits >= 32) {
+			alignBits = 32;
+			alignMask = 0;
+		}
+		else {
+			alignMask = ~((1u << alignBits) - 1);
+		}
+
+		*allocSize = 0;
+		srBase = addr & alignMask;
+		srSize = 1u << (alignBits - 3);
+
+		for (bit = 0; bit < 8; bit++) {
+			if (srBase < addr || srBase + srSize > addr + size || *allocSize + srSize > size)
+				srdMask |= 1u << bit;
+			else
+				*allocSize += srSize;
+
+			srBase += srSize;
+		}
+	}
+
+	return mpu_regionSet(region, addr & alignMask, alignBits, attr, enable, srdMask);
+}
+
+
+/* Allocate MPU sub/regions for a map */
+static int mpu_regionAlloc(u8 *regCnt, u32 mapId, u32 addr, u32 size, u32 attr, u8 enable)
+{
+	u32 allocSize = 0;
+	u8 regCur = *regCnt;
+
+	while (size > 0 && (regCur - *regCnt) < 2) {
+		if (mpu_regionBestFit(regCur, addr, size, attr, enable, &allocSize) < 0)
+			return -1;
+
+		if (allocSize > size)
+			return -1;
+
+		addr += allocSize;
+		size -= allocSize;
+		regCur++;
+	}
+
+	if (size != 0)
+		return -1;
+
+	for (;; (*regCnt)++) {
```

Comment:
do while fits here better:

```
do
 mpu_common.hal.mpu.map[(*regCnt)++] = mapId;
while (*regCnt != regCur);
```

Reply:
If You meant this:
```
    do {
         mpu_common.hal.mpu.map[*regCnt] = mapId;
    } while (*regCnt != regCur && ++(*regCnt));

```
that's ok, but it is not MISRA conformant because of ++ in `while` loop, thus more readable is current version.

Reply:
Why not `mpu_common.hal.mpu.map[(*regCnt)++] = mapId;`? Then you don't have `++` in `while`.

Reply:
Don't want to increment `regCnt` if `*regCnt >= regCur`, maybe the updated version is more clearer.

Reply:
Wouldn't wersion with do..while achieve the same result?

Reply:
Ok, I get it, so do..while doesn't fit that well afterall